src/content/filtering-string.js

```diff
@@ -0,0 +1,141 @@
+// @flow
+
+import isEqual from 'lodash.isequal';
+
+import type { IndexIntoStackTable } from '../common/types/profile';
+import type { ImplementationFilter } from './actions/types';
+
+export type FilterDescription = {|
+  implementation: ImplementationFilter,
+  substrings: string[],
+  paths: string[],
+|};
+
+export type ChargeDescription = {|
+  source: IndexIntoStackTable,
+  target: IndexIntoStackTable,
+|};
+
+export type Filter = {|
+  include: FilterDescription | null,
+  exclude: FilterDescription | null,
+  display: {|
+    invertCallstack: boolean,
+    hidePlatformDetails: boolean,
+    hide: IndexIntoStackTable[],
+    charge: ChargeDescription[],
+  |},
+  cachedString: string,
+|};
+
+function _fromFilterDescription(filter: FilterDescription): string[] {
+  const result = [];
+
+  if (filter.implementation) {
+    result.push(`implementation:${filter.implementation}`);
+  }
+
+  result.push(...filter.paths.map(path => `path:${path}`));
+  result.push(...filter.substrings.map(string => `substring:${string}`));
+
+  return result;
+}
+
+export function stringFromFilter(filter: Filter | null = null): string {
+  if (!filter) {
+    return '';
+  }
+
+  if (isEqual(filter, filterFromString(filter.cachedString))) {
+    return filter.cachedString;
+  }
+
+  const result = [];
+  if (filter.include) {
+    result.push(..._fromFilterDescription(filter.include));
+  }
+
+  if (filter.exclude) {
+    const excludedItems = _fromFilterDescription(filter.exclude).map(item => `-${item}`);
+    result.push(...excludedItems);
+  }
+
+  const display = filter.display;
+  if (display.invertCallstack) {
+    result.push('invertCallstack');
+  }
+  if (display.hidePlatformDetails) {
+    result.push('hidePlatformDetails');
+  }
+  result.push(...display.hide.map(index => `hide:${index}`));
+  result.push(...display.charge.map(({ source, target }) => `charge:${source}:${target}`));
+
+  return result.join(' ');
+}
+
+export function emptyUserFilter(): Filter {
+  return {
+    include: null,
+    exclude: null,
+    display: {
+      invertCallstack: false,
+      hidePlatformDetails: false,
+      hide: [],
+      charge: [],
+    },
+    cachedString: '',
+  };
+}
+
+const reFilterFromString = /([-+])?\b(\w+)(?::(\S+))?(?=\s|$)/g;
```

Comment:
Holy regex Batman! Would probably be good to document this for the final. Also I'm wondering if a regex will scale and be maintainable compared to a parser.